spark-atlas-connector/src/main/scala/com/hortonworks/spark/atlas/AtlasEntityCreationRequestHelper.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hortonworks.spark.atlas
+
+import java.util.UUID
+
+import com.hortonworks.spark.atlas.types.metadata
+import com.hortonworks.spark.atlas.utils.Logging
+import org.apache.atlas.model.instance.AtlasObjectId
+
+import scala.collection.mutable
+
+class AtlasEntityCreationRequestHelper(
+    atlasClient: AtlasClient) extends Logging {
+  // query to (inputs, outputs)
+  private val queryToInputsAndOutputs = new mutable.HashMap[UUID,
+    (Set[AtlasObjectId], Set[AtlasObjectId])]()
+
+  def requestCreation(entities: Seq[SACAtlasReferenceable], runId: Option[UUID] = None): Unit = {
+    runId match {
+      case Some(rid) => updateEntitiesForStreamingQuery(rid, entities)
+      case None => updateEntitiesForBatchQuery(entities)
+    }
+  }
+
+  private def updateEntitiesForBatchQuery(entities: Seq[SACAtlasReferenceable]): Unit = {
+    // the query is batch, hence always create entities
+    // create input/output entities as well as update process entity(-ies)
+    createEntities(entities)
+  }
+
+  private def updateEntitiesForStreamingQuery(
+      runId: UUID,
+      entities: Seq[SACAtlasReferenceable]): Unit = {
+    // the query is streaming, so which partial of source/sink entities can be seen
+    // in specific batch - need to accumulate efficiently
+    val processes = entities
+      .filter(en => en.typeName == metadata.PROCESS_TYPE_STRING
+        && en.isInstanceOf[SACAtlasEntityWithDependencies])
+      .map(_.asInstanceOf[SACAtlasEntityWithDependencies])
+
+    val inputs = processes.flatMap { p =>
+      AtlasEntityReadHelper.getSeqAtlasObjectIdAttribute(p.entity, "inputs")
+    }.toSet
+
+    val outputs = processes.flatMap { p =>
+      AtlasEntityReadHelper.getSeqAtlasObjectIdAttribute(p.entity, "outputs")
+    }.toSet
+
+    queryToInputsAndOutputs.get(runId) match {
```

Comment:
The `runId` can change if a query is stopped and re-started within same spark application. Since we want to track the lineage based on application Id, we don't want to re-create the lineage if a query is just restarted within same application correct? If so the key should be query-id here.

Reply:
Nice finding. Will address shortly.